### PR TITLE
Allow calculation of dihedral angles

### PR DESCRIPTION
**Pull request by jbarnoud**
Add a function to calculate dihedral angles, but also dihedral angles shifted by -pi as used by scFix. The function also adds the corresponding link parameter effectors.

The function to calculate the angles is moved to the new `geometry.py` module that even has tests.

### CODE REVIEW

File: vermouth/geometry.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    if cosine > 1:
+        cosine = 1
+    elif cosine < -1:
+        cosine = -1
+    return np.arccos(cosine)
+
+
+def dihedral(vectorAB, vectorBC, vectorCD):
+    """
+    Calculate the dihedral angle in radians formed by 3 vectors.
```
Comment (pckroon) [inline vermouth/geometry.py]: By the strict definition an (improper) dihedral is the angle between 2 planes, each spanned by 2 vectors. I would change the signature to take either two pairs of vectors, or 4 vectors.
What you're describing here is a torsion angle.

File: vermouth/geometry.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        A - B
+             \
+              C - D
+    
+    The angle around the BC axis is returned with values between -pi and +pi.
+    The direction of the angle follows the right hand convention.
+    """
+    normalABC = np.cross(vectorAB, vectorBC)
+    normalBCD = np.cross(vectorBC, vectorCD)
```
Comment (pckroon) [inline vermouth/geometry.py]: Do these need to be normalized?
What happens if ABC or BCD is (nearly) colinear?

File: vermouth/geometry.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    psin = np.dot(normalABC, vectorCD) * np.linalg.norm(vectorBC)
+    pcos = np.dot(normalABC, normalBCD)
+    return np.arctan2(psin, pcos)
+
+
+def dihedral_phase(vectorAB, vectorBC, vectorCD):
+    """
+    Calculate a dihedral angle in radians with a -pi phase correction.
```
Comment (pckroon) [inline vermouth/geometry.py]: I get the feeling the phase correction should be an argument

File: vermouth/geometry.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    --------
+    dihedral
+        Calculate a dihedral angle.
+    """
+    angle = dihedral(vectorAB, vectorBC, vectorCD)
+    angle -= np.pi
+    if angle > np.pi:
+        angle -= 2 * np.pi
+    if angle < -np.pi:
+        angle += 2 * np.pi
```
Comment (pckroon) [inline vermouth/geometry.py]: And if it is an argument, these should be while statements.

File: vermouth/tests/test_geometry.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        The angle between the 3 points.
+    """
+    # The shift is arbitrary. Its purpose is to avoid having the angle centered
+    # at the origin as it may be a special case.
+    shift = np.array([1.1, -7.2, 9.1])
+    coordinates = np.array([
+        [2, 0, 0],
+        [0, 0, 0],
+        [0, 0, 0],  # Will be redifined
+    ]) + shift
```
Comment (pckroon) [inline vermouth/tests/test_geometry.py]: `np.full((3, 3), shift)`; `coords[0,0]+=2`?

File: vermouth/tests/test_geometry.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    Yields
+    ------
+    coordinates: np.ndrarray
+        The coordinates of the points. Each row corresponds to a pointm and
+        each column corresponds to a dimension.
+    angle: float
+        The angle between the 3 points.
+    """
+    # The shift is arbitrary. Its purpose is to avoid having the angle centered
+    # at the origin as it may be a special case.
```
Comment (pckroon) [inline vermouth/tests/test_geometry.py]: So make sure we also test the special case ;)

File: vermouth/tests/test_geometry.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    shift = np.array([1.1, -7.2, 9.1])
+    coordinates = np.array([
+        [2, 0, 0],
+        [0, 0, 0],
+        [0, 0, 5],
+        [0, 0, 5],  # Will be redifined
+    ]) + shift
```
Comment (pckroon) [inline vermouth/tests/test_geometry.py]: Same comments as above

File: vermouth/geometry.py (hunk 3 above)
Comment (jbarnoud) [reply vermouth/geometry.py]: It seems that an phase of -pi is a specific and useful case. It seems to be use by gromacs. I fail to grasp why, though. I'll add later a function that can rotate any angle and keep it in bound. In an other PR if you do not mind.

File: vermouth/tests/test_geometry.py (hunk 6 above)
Comment (pckroon) [reply vermouth/tests/test_geometry.py]: I understand that. By the way, are you sure the broadcasting of (3, 3) + (3, ) happens as you expect it to? Not that I really matters if it doesn't.

File: vermouth/geometry.py (hunk 3 above)
Comment (pckroon) [reply vermouth/geometry.py]: Sure. Just don't forget ;)

File: vermouth/tests/test_geometry.py [hunk 8]
```diff
@@ -101,29 +102,56 @@ def _generate_test_dihedrals(n_angles):
...
-    assert np.allclose(geometry.dihedral(vectorAB, vectorBC, vectorCD), angle)
-
-
-@pytest.mark.parametrize('points, angle', _generate_test_dihedrals(10))
+    calc_angle = geometry.dihedral(points)
+    # +pi and -pi are the same angle; we normalize them to pi
+    if np.allclose(calc_angle, -np.pi):
+        calc_angle *= -1
+    if np.allclose(angle, -np.pi):
+        angle *= -1
```
Comment (pckroon) [inline vermouth/tests/test_geometry.py]: I get the feeling this is not entirely correct. Adding 2 pi would be better I think.

File: vermouth/tests/test_geometry.py [hunk 9]
```diff
@@ -101,29 +102,56 @@ def _generate_test_dihedrals(n_angles):
...
         angle_phase -= 2 * np.pi
     if angle_phase < -np.pi:
         angle_phase += 2 * np.pi
-    assert np.allclose(geometry.dihedral_phase(vectorAB, vectorBC, vectorCD), angle_phase)
+    calc_angle = geometry.dihedral_phase(points)
+    # +pi and -pi are the same angle; we normalize them to pi
+    if np.allclose(calc_angle, -np.pi):
+        calc_angle *= -1
+    if np.allclose(angle_phase, -np.pi):
+        angle_phase *= -1
```
Comment (pckroon) [inline vermouth/tests/test_geometry.py]: Same as above

File: vermouth/geometry.py [hunk 10]
```diff
@@ -36,43 +36,56 @@ def angle(vectorBA, vectorBC):
...
-    if cosine > 1:
-        cosine = 1
-    elif cosine < -1:
-        cosine = -1
+    cosine = np.clip(cosine, -1, 1)
     return np.arccos(cosine)
 
 
-def dihedral(vectorAB, vectorBC, vectorCD):
+def dihedral(coordinates):
```
Comment (pckroon) [inline vermouth/geometry.py]: I liked the previous signature better. But if you argue this is a low-level function, sure. Just make sure there's also human usable ones which calculate torsion angles and angles between planes respectively.

File: vermouth/tests/test_geometry.py (hunk 8 above)
Comment (jbarnoud) [reply vermouth/tests/test_geometry.py]: It should be the exact same, I just want -pi and pi to compare to the same thing.

File: vermouth/geometry.py (hunk 10 above)
Comment (jbarnoud) [reply vermouth/geometry.py]: In practice I always used all dihedral calculation functions with 4 points. It is what gromacs, MDAnalysis, or VMD do. I could have a 6 points version, or a 2 times 3 point version as well.

Comment (pckroon) [reply vermouth/geometry.py]: The problem with a 4pt version is that impropers and torsion angles become separate things. I do agree it's the most common thing you'll find.

**Pull request merged**